src/main/scala/zio/prelude/NonEmptySet.scala

```diff
@@ -0,0 +1,207 @@
+
+  /**
+   * Converts this `NonEmptySet` to a `NonEmptyList`.
+   */
+  def toNonEmptyList: NonEmptyList[A] = destruct match { case (head, tail) => NonEmptyList.fromIterable(head, tail) }
+
+  /** Creates a new `NonEmptySet` with an additional element, unless the element is
+   *  already present.
+   *
+   *  @param elem the element to be added
+   *  @return a new set that contains all elements of this set and that also
+   *          contains `elem`.
+   */
+  def +(elem: A): NonEmptySet[A] = new NonEmptySet(toSet + elem)
+
+  /** Creates a new `NonEmptySet` with additional elements, omitting duplicates.
+   *
+   *  This method takes two or more elements to be added. Elements that already exist in the `NonEmptySet` will
+   *  not be added. Another overloaded variant of this method handles the case where a single element is added.
+   *
+   *  Example:
+   *   {{{
+   *    scala> val a = NonEmptySet(1, 3) + 2 + 3
+   *    a: zio.prelude.NonEmptySet[Int] = NonEmptySet(1, 3, 2)
+   *   }}}
+   *
+   *  @param elem1 the first element to add.
+   *  @param elem2 the second element to add.
+   *  @param elems the remaining elements to add.
+   *  @return   a new `NonEmptySet` with the given elements added, omitting duplicates.
+   */
+  def +(elem1: A, elem2: A, elems: A*): NonEmptySet[A] = new NonEmptySet(toSet + elem1 + elem2 ++ elems)
+
+  /** Computes the union between of `NonEmptySet` and another set.
+   *
+   *  @param   that  the set to form the union with.
+   *  @return  a new `NonEmptySet` consisting of all elements that are in this
+   *  set or in the given set `that`.
+   */
+  def union(that: Set[A]): NonEmptySet[A] = new NonEmptySet(toSet.union(that))
+
+  /** Creates a new `NonEmptySet` by adding all elements contained in another collection to this `NonEmptySet`, omitting duplicates.
+   *
+   * This method takes a collection of elements and adds all elements, omitting duplicates, into `NonEmptySet`.
+   *
+   * Example:
+   *  {{{
+   *    scala> val a = NonEmptySet(1, 2) ++ NonEmptySet(2, "a")
+   *    a: zio.prelude.NonEmptySet[Any] = NonEmptySet(1, 2, a)
+   *  }}}
+   *
+   *  @param elems     the collection containing the elements to add.
+   *  @return a new `NonEmptySet` with the given elements added, omitting duplicates.
+   */
+  def ++(elems: Iterable[A]): NonEmptySet[A] = new NonEmptySet(toSet ++ elems)
+
+  /** Adds the `elem` to this `NonEmptySet`. Alias for `+`. */
+  def add(elem: A): NonEmptySet[A] = self + elem
+
+  /** Removes the `elem` from this `NonEmptySet`. Alias for `-`. */
+  def remove(elem: A): Set[A] = toSet - elem
+
+  /**
+   * Flattens a `NonEmptySet` of `NonEmptySet` values into a single
+   * `NonEmptySet`.
+   */
+  def flatten[B](implicit ev: A <:< NonEmptySet[B]): NonEmptySet[B] =
+    destruct match {
+      case (head, tail) =>
+        tail.foldLeft(ev(head)) { (b, a) =>
+          b ++ ev(a)
+        }
+    }
+
+  override def hashCode: Int = toSet.hashCode()
+
```

Comment:
```suggestion
set.foldLeft[Set[B]](Set.empty)((b, a) => b union ev(a))
```
I think slightly more efficient because avoids destructuring the set.